File: day6/part1.ts
import * as fs from "fs";

function readInput(path: string): string[] {
  return fs.readFileSync(path, "utf8").split("\n");
}

function splitBlocks(lines: string[]): string[][] {
  const width = Math.max(...lines.map((l) => l.length));
  const padded = lines.map((l) => l.padEnd(width, " "));

  const used = Array(width).fill(false);
  for (let c = 0; c < width; c++) {
    for (const row of padded) {
      if (row[c] !== " ") {
        used[c] = true;
        break;
      }
    }
  }

  const blocks: string[][] = [];
  let start: number | null = null;

  for (let i = 0; i <= width; i++) {
    const active = i < width && used[i];
    if (active && start === null) {
      start = i;
    } else if (!active && start !== null) {
      blocks.push(padded.map((r) => r.slice(start!, i)));
      start = null;
    }
  }

  return blocks;
}

function solveBlock(block: string[]): bigint {
  const numbers: bigint[] = [];
  let op: string | null = null;

  for (const row of block) {
    const s = row.trim();
    if (!s) continue;
    if (s === "+" || s === "*") {
      op = s;
    } else {
      numbers.push(BigInt(s));
    }
  }

  if (op === "+") {
    return numbers.reduce((a, b) => a + b, 0n);
  }
  return numbers.reduce((a, b) => a * b, 1n);
}

const lines = readInput("input.txt");
const blocks = splitBlocks(lines);
const total = blocks.reduce((sum, b) => sum + solveBlock(b), 0n);

console.log(total.toString());

File: day6/part1.py
from typing import List
import math


def read_input(path: str) -> List[str]:
    with open(path, "r", encoding="utf-8") as f:
        return [line.rstrip("\n") for line in f]


def split_blocks(lines: List[str]) -> List[List[str]]:
    width = max(len(line) for line in lines)
    padded = [line.ljust(width) for line in lines]

    used = [any(row[c] != " " for row in padded) for c in range(width)]

    blocks: List[List[str]] = []
    start: int | None = None

    for i, active in enumerate(used + [False]):
        if active and start is None:
            start = i
        elif not active and start is not None:
            blocks.append([row[start:i] for row in padded])
            start = None

    return blocks


def solve_block(block: List[str]) -> int:
    numbers: List[int] = []
    operator: str | None = None

    for row in block:
        s = row.strip()
        if not s:
            continue
        if s in ("+", "*"):
            operator = s
        else:
            numbers.append(int(s))

    if operator == "+":
        return sum(numbers)
    if operator == "*":
        return math.prod(numbers)

    raise ValueError("Operator not found in block")


def main() -> None:
    lines = read_input("input.txt")
    blocks = split_blocks(lines)
    total = sum(solve_block(b) for b in blocks)
    print(total)


if __name__ == "__main__":
    main()
